### DISCUSSION TITLE
invalidateDecorationsAllViews - not working?

### DISCUSSION THREAD

**Discussion by Johannes-repos** [iTwin/itwinjs-core#1511] [Q&A] [answered]
I am trying to link external sensor data from a Home Assistant https://www.home-assistant.io/ / https://www.npmjs.com/package/homeassistant instance into the viewer app.
Data is coming in fine and is displayed by the decorators.

![image](https://user-images.githubusercontent.com/39602714/119641371-7a451000-be11-11eb-80e7-0dc6909c0d6f.png)

 
Values for this sensor are updated every 10 seconds.

![2021-05-26_11-03-56](https://user-images.githubusercontent.com/39602714/119641947-1c64f800-be12-11eb-8420-ad181dfe7986.png)

To update the live data feed, I am  trying to use:

`

      setInterval( () => {
        const d=new SmartDeviceDecorator(vp);
        console.log(IModelApp.viewManager.dropDecorator(d));
        IModelApp.viewManager.invalidateDecorationsAllViews();
        IModelApp.viewManager.invalidateCachedDecorationsAllViews(d);
        IModelApp.viewManager.addDecorator(d);

        vp.invalidateDecorations(); // Refresh the markers on the screen
      }, 5000);
 `

Instead of updating or removing the decorators every 5 seconds, a new decorator is added. 
As visible above, the sun icon gets more and more blurry.

Here’s a link to a video describing the issue:

https://www.dropbox.com/s/o8bbna3co27xiii/invalidateDecorationsAllViews.mp4?dl=0

Any idea how I could resolve this?

Thanks

**Comment by pmconne** [comment, marked as answer]
Based on your code snippet, you create and add a new SmartDeviceDecorator every 5 seconds, and never drop the previously-added ones.
Try something like:
```
setInterval(() => {
  const oldDecorator = IModelApp.viewManager.decorators.find((dec) => dec instanceof SmartDeviceDecorator);
  if (oldDecorator)
    IModelApp.viewManager.dropDecorator(oldDecorator);

  IModelApp.viewManager.addDecorator(new SmartDeviceDecorator(vp));
}, 5000);
```
Alternatively, if there's only ever one SmartDeviceDecorator, just store the most recent one in some variable instead of searching for it.

Note: add/dropDecorator take care of invalidating decorations - you should not have to do that yourself here.

**Reply by Johannes-repos** [reply]
Hi Paul,

thanks for your quick reply and you're solution worked perfectly!
